api/fbs/v1/query.fbs

```diff
@@ -0,0 +1,138 @@
+// specific language governing permissions and limitations
+// under the License.
+
+namespace banyandb.v1;
+
+// BinaryOp specify the operation imposed to the given query condition
+enum BinaryOp: byte {
+    EQ,
+    NE,
+    LT,
+    GT,
+    LE,
+    GE,
+    HAVING,
+    NOT_HAVING
+}
+
+// BinaryOps is an array of binary operators.
+table BinaryOps {
+    ops: [BinaryOp];
+}
+
+// IntPair in a typed pair with an array of int64 as values
+table IntPair {
+    key: string;
+    values: [int64];
+}
+
+// StrPair in a typed pair with an array of string as values
+table StrPair {
+    key: string;
+    values: [string];
+}
+
+// Pair is the building block of a record which is equivalent to a key-value pair.
+// In the context of Trace, it could be metadata of a trace such as service_name, serivce_instance, etc.
+// Besides, other fields/tags are organized in key-value pair in the underlying storage layer.
+// One should notice that the values can be a multi-value.
+union TypedPair { IntPair, StrPair }
+
+// Pair is a workaround since vector of unions is not yet supported in some of languages.
+table Pair {
+    pair: TypedPair;
+}
+
+// PairQuery consists of the query condition with a binary operator to be imposed
+table PairQuery {
+    ops: BinaryOps;
+    condition: Pair;
+}
+
+// Sort is either descending or ascending
+enum Sort : byte {
+    DESC,
+    ASC
+}
+
+// QueryOrder means a Sort operation to be done for a given field.
+// The key_name refers to the key of a Pair.
+table QueryOrder {
+    key_name: string;
+    sort: Sort;
+}
+
+// Entity represents 
+// (Trace context) a Span defined in Google Dapper paper or equivalently a Segment in Skywalking.
+// (Log context) a log
+table Entity {
+    // entity_id could be span_id of a Span or segment_id of a Segment in the context of Trace
+    entity_id: string;
+    // timestamp_nanoseconds is in the timeunit of nanoseconds. It represents
+    // 1) either the start time of a Span/Segment,
+    // 2) or the timestamp of a log
+    timestamp_nanoseconds: uint64;
+    // data_binary contains all unindexed Tags and other key-value pairs
+    data_binary: [ubyte];
+    // fields contains all indexed Field. Some typical names,
+    // - trace_id
+    // - duration
+    // - service_name
+    // - service_instance_id
+    // - end_time_nanoseconds
+    fields: [Pair];
+}
+
+// TracesResponse is the response for a query to the Query module. 
+table TracesResponse {
+    // entities are the actual data returned
+    entities: [Entity];
+}
+
+// Projection is used to select the names of keys to be returned.
+table Projection {
+    // The key_name refers to the key(s) of Pair(s).
+    key_names: [string];
+}
+
+// RangeQuery is a range query for uint64,
+// the range here follows left-inclusive and right-exclusive rule, i.e. [begin, end) if both edges exist
+table RangeQuery {
+    begin: uint64;
+    end: uint64;
+}
+
+// EntityCriteria is the request contract for query.
+table EntityCriteria {
```

Comment:
Once this is a query/insert/update parameter, how about we put a `metadata` field in it for the schema/tenancy/authentification extension?

Reply:
yes, in the logical model design, trace series should have a `name`, and be under a `group`.
https://docs.google.com/document/d/1XFzHWXSWyIC7AVNIx0ifQ0eJAJYfoPO-2j_lgbl9zu8/edit#heading=h.qr4os6u8z3fu

Reply:
Metadata should be reused in modification operations.

Reply:
Group and name is the thing you have to carry everytime. Don't they?

Reply:
I've added metadata to `database.fbs`

Reply:
Write protocol might encode them because it’s streaming. The query doesn’t have such concerns

Reply:
> Write protocol might encode them because it’s streaming. The query doesn’t have such concerns

Yes, streaming only means metadata exists in the first package. From the protocol(fbs) perspective, it is the same.

Reply:
Make sense to me